functions/src/aqi-calculation/purple-air-response.ts

```diff
@@ -1,32 +1,370 @@
+      return NaN;
+    case maxConfidence:
+      // The confidence value from PurpleAir can be 100 even if channel A and
+      // channel B do not completely match, but if the confidence value is 100,
+      // then the value is data is good enough to meet the EPA recommendation.
+      return 0;
+    default:
+      // Otherwise, undo the calculation from the PurpleAir confidence value
+      /* eslint-disable-next-line no-magic-numbers */
+      return ((maxConfidence - confidence + 25) * 1.6) / 100;
+  }
+}
+
+/**
+ * Converts a PurpleAir reading returned from the group API query into a PurpleAirReading
+ * @param data - list of data from PurpleAir for a given sensor
+ * @param fieldNames - list of field names from PurpleAir that match the order of the data fields
+ * @returns
+ */
+function getReading(
+  data: (string | number)[],
+  fieldNames: string[]
+): [number, PurpleAirReading | null] {
+  // Initialize all values
+  let id: number = Number.NaN;
+  let name: string | undefined = undefined;
+  let latitude: number | undefined = undefined;
+  let longitude: number | undefined = undefined;
+  let meanPercentDifference: number | undefined = undefined;
+  let pm25: number | undefined = undefined;
+  let humidity: number | undefined = undefined;
+  let timestamp: Date | undefined = undefined;
+
+  data.forEach((value, index) => {
+    // Check the corresponding field name to determine how to handle the value
+    switch (fieldNames[index]) {
+      case 'sensor_index':
+        if (typeof value === 'number') id = value;
+        break;
+      case 'name':
+        if (typeof value === 'string') name = value;
+        break;
+      case 'latitude': {
+        if (typeof value === 'number') latitude = value;
+        break;
+      }
+      case 'longitude':
+        if (typeof value === 'number') longitude = value;
+        break;
+      case 'confidence':
+        if (typeof value === 'number') {
+          meanPercentDifference = getMeanPercentDifference(value);
+        }
+        break;
+      case 'pm2.5':
+        if (typeof value === 'number') pm25 = value;
+        break;
+      case 'humidity':
+        if (typeof value === 'number') humidity = value;
+        break;
+      case 'last_seen':
+        if (typeof value === 'number') {
+          // PurpleAir returns seconds since EPOCH, but the Date constructor
+          // takes milliseconds, so we convert from seconds to milliseconds
+          timestamp = new Date(value * 1000); // eslint-disable-line no-magic-numbers
+        }
+        break;
+      default:
+        // Unknown field, ignore
+        break;
+    }
+  });
+
+  // Only return a PurpleAirReading if all fields are defined
+  if (
+    id !== undefined &&
+    name !== undefined &&
+    latitude !== undefined &&
+    longitude !== undefined &&
+    meanPercentDifference !== undefined &&
+    pm25 !== undefined &&
+    humidity !== undefined &&
+    timestamp !== undefined
```

Comment:
Can't we just use `id && name && etc..` because undefined is a falsey value? Or do we want to make sure that nulls are still okay?

Reply:
I check undefined specifically because 0s are okay, which is also a falsey value.